src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -220,40 +323,380 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
                                 exceptions.Add(ex);
                             }
                         }
+
+                        if (reversedLifetimeServices is not null)
+                        {
+                            foreach (IHostedLifecycleService hostedService in reversedLifetimeServices)
+                            {
+                                try
+                                {
+                                    await hostedService.StoppedAsync(token).ConfigureAwait(false);
+                                }
+                                catch (Exception ex)
+                                {
+                                    exceptions.Add(ex);
+                                }
+                            }
+                        }
                     }
                 }
 
+                LogExceptions();
+
                 // Fire IHostApplicationLifetime.Stopped
                 _applicationLifetime.NotifyStopped();
 
+                void LogExceptions()
+                {
+                    if (exceptions.Count > 0)
+                    {
+                        if (exceptions.Count == 1)
+                        {
+                            // Rethrow if it's a single error
+                            Exception singleException = exceptions[0];
+                            _logger.StoppedWithException(singleException);
+                            ExceptionDispatchInfo.Capture(singleException).Throw();
+                        }
+                        else
+                        {
+                            var ex = new AggregateException("One or more hosted services failed to stop.", exceptions);
+                            _logger.StoppedWithException(ex);
+                            throw ex;
+                        }
+                    }
+                }
+            }
+
+            _logger.Stopped();
+        }
+
+        /// <summary>
+        /// Call <see cref="IHostedLifecycleService.StartingAsync(CancellationToken)"/>.
+        /// The beginning synchronous portions of the implementations are run serially in registration order for
+        /// performance since it is common to return Task.Completed as a noop.
+        /// Any subsequent asynchronous portions are grouped together run concurrently.
+        /// </summary>
+        private static async Task CallLifeCycle_Starting(IEnumerable<IHostedLifecycleService> services, CancellationToken token, List<Exception> exceptions)
```

Comment:
Could we avoid so much duplication by having a method that allows the caller to specify concurrency and what operation to do on each service? 
```c#
private static async Task ForeachService<T>(IEnumerable<T> services, CancellationToken token, List<Exceptions> exceptions, bool concurrent, Func<T, CancellationToken, Task> operation)
```

Reply:
The latest commit does this, but had to also pass in some extra state. I wasn't sure if the extra level of indirection was worth it - PTAL.

Also, a couple newer tests broke which were created when concurrent startup was added. The tests assumed an InvalidOperationException would be thrown in concurrent case + one exception, but now there is an AggregateException (with the one inner exception). The extra level of Task indirection apparently caused that, and in general I believe AggregateException is expected here.

Reply:
Thanks - the latest commit does this. I wasn't sure if the extra delegate hop and use of generics was worth it, but it does reduce the code size.